### ISSUE TITLE
Error happened when using pika module

### ISSUE THREAD

**Issue by deliciousdish** [gramineproject/gramine#920]
## Description of the problem
Error happened when using pika module， which was used to connect a message queue.

## Steps to reproduce
First, install pika module.
```
sudo apt install python3-pika
sudo apt install rabbitmq-server
```
Second, modify helloworld.py
```
#!/usr/bin/env python3
import pika

connection = pika.BlockingConnection(pika.ConnectionParameters(host = 'localhost'))
```
## Actual results
```
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - sgx.debug = true                           (this is a debug enclave)
  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)
  - sgx.allowed_files = [ ... ]                (some files are passed through from untrusted host without verification)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

[P1:T1:python3.8] error: Internal memory fault at 0x00000000 (0x1d93c993, VMID = 1, TID = 1)
```
I can't upload logs, so i paste the last part of logs here.
```
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/tasks.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x164353f0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16435450) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164357a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x158941f0, 0x8484) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x8483
[P1:T1:python3.8] trace: ---- read(3, 0x1589c673, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/tasks.py", 0x16435510) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16435090) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/tasks.cpython-38.pyc.357504032'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/tasks.cpython-38.pyc.357504032", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x164367c0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/trsock.py", 0x164364a0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/trsock.py", 0x16436ca0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/trsock.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x164369f0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16436a50) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x16436da0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x1554cb20, 0x2124) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x2123
[P1:T1:python3.8] trace: ---- read(3, 0x1554ec43, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/trsock.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x164369f0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16436a50) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x16436da0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x1553b3a0, 0x16f5) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x16f4
[P1:T1:python3.8] trace: ---- read(3, 0x1553ca94, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/trsock.py", 0x16436b10) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16436690) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/trsock.cpython-38.pyc.361799440'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/trsock.cpython-38.pyc.361799440", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x16437dc0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/runners.py", 0x16437aa0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/runners.py", 0x164382a0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/runners.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x150a0810, 0x79c) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x79b
[P1:T1:python3.8] trace: ---- read(3, 0x150a0fab, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/runners.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x1509fad0, 0x807) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x806
[P1:T1:python3.8] trace: ---- read(3, 0x150a02d6, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/runners.py", 0x16438110) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16437c90) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/runners.cpython-38.pyc.349581360'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/runners.cpython-38.pyc.349581360", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x16437dc0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/queues.py", 0x16437aa0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/queues.py", 0x164382a0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/queues.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x1553b3a0, 0x20be) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x20bd
[P1:T1:python3.8] trace: ---- read(3, 0x1553d45d, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/queues.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x1553d490, 0x2020) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x201f
[P1:T1:python3.8] trace: ---- read(3, 0x1553f4af, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/queues.py", 0x16438110) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16437c90) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/queues.cpython-38.pyc.349581360'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/queues.cpython-38.pyc.349581360", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x16437dc0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/streams.py", 0x16437aa0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/streams.py", 0x164382a0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/streams.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x1553b3a0, 0x50a1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x50a0
[P1:T1:python3.8] trace: ---- read(3, 0x15540440, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/streams.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x15095a10, 0x6821) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x6820
[P1:T1:python3.8] trace: ---- read(3, 0x1509c230, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/streams.py", 0x16438110) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16437c90) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/streams.cpython-38.pyc.361798992'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/streams.cpython-38.pyc.361798992", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x16437dc0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/subprocess.py", 0x16437aa0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/subprocess.py", 0x164382a0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/subprocess.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x15544130, 0x1cbf) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x1cbe
[P1:T1:python3.8] trace: ---- read(3, 0x15545dee, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/subprocess.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x15545e20, 0x1f85) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x1f84
[P1:T1:python3.8] trace: ---- read(3, 0x15547da4, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/subprocess.py", 0x16438110) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16437c90) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/subprocess.cpython-38.pyc.349490480'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/subprocess.cpython-38.pyc.349490480", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x16437dc0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/unix_events.py", 0x16437aa0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/unix_events.py", 0x164382a0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/unix_events.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x158a5a30, 0x99a9) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x99a8
[P1:T1:python3.8] trace: ---- read(3, 0x158af3d8, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/unix_events.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x16437ff0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16438050) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x164383a0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x15529700, 0xbfdc) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0xbfdb
[P1:T1:python3.8] trace: ---- read(3, 0x155356db, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/unix_events.py", 0x16438110) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16437c90) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/unix_events.cpython-38.pyc.349490480'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/unix_events.cpython-38.pyc.349490480", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x164367c0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/base_subprocess.py", 0x164364a0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/base_subprocess.py", 0x16436ca0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/base_subprocess.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x164369f0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16436a50) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x16436da0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x15091020, 0x24c9) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x24c8
[P1:T1:python3.8] trace: ---- read(3, 0x150934e8, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/base_subprocess.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x164369f0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16436a50) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x16436da0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x15544ed0, 0x228c) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x228b
[P1:T1:python3.8] trace: ---- read(3, 0x1554715b, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/base_subprocess.py", 0x16436b10) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16436690) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/base_subprocess.cpython-38.pyc.358605104'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/base_subprocess.cpython-38.pyc.358605104", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio", 0x164367c0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/selector_events.py", 0x164364a0) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/selector_events.py", 0x16436ca0) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/selector_events.cpython-38.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x164369f0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16436a50) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x16436da0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x1553b3a0, 0x73f6) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x73f5
[P1:T1:python3.8] trace: ---- read(3, 0x15542795, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/selector_events.py", O_RDONLY|0x80000, 0000) = 0x3
[P1:T1:python3.8] trace: ---- fstat(3, 0x164369f0) = 0x0
[P1:T1:python3.8] trace: ---- ioctl(3, TCGETS, 0x16436a50) ...
[P1:T1:python3.8] trace: ---- return from ioctl(...) = -38
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- lseek(3, 0x0, SEEK_CUR) = 0x0
[P1:T1:python3.8] trace: ---- fstat(3, 0x16436da0) = 0x0
[P1:T1:python3.8] trace: ---- read(3, 0x15085030, 0x9855) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x9854
[P1:T1:python3.8] trace: ---- read(3, 0x1508e884, 0x1) ...
[P1:T1:python3.8] trace: ---- return from read(...) = 0x0
[P1:T1:python3.8] trace: ---- close(3) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/selector_events.py", 0x16436b10) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x16436690) = 0x0
warning: Disallowing access to file '/usr/lib/python3.8/asyncio/__pycache__/selector_events.cpython-38.pyc.358605104'; file is not trusted or allowed.
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/selector_events.cpython-38.pyc.358605104", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- clock_gettime(0, 0x1643b5b0) = 0x0
[P1:T1:python3.8] trace: ---- epoll_create1(524288) = 0x3
[P1:T1:python3.8] trace: ---- clock_gettime(0, 0x1643bc90) = 0x0
[P1:T1:python3.8] trace: ---- clock_gettime(1, 0x1643bde0) = 0x0
[P1:T1:python3.8] trace: ---- epoll_wait(3, 0x15544710, 1023, 0) ...
[P1:T1:python3.8] error: Internal memory fault at 0x00000000 (0x1d18e793, VMID = 1, TID = 1)
debug: PalProcessExit: Returning exit code 1
```
## Gramine commit hash
07fb01664bf692e851792866cc81db32636818f3

**Comment by deliciousdish** [comment]
gramine-direct also failed.
```
caiqishen@caiqishen-ThinkCentre-M920t-N000:~/gramine/CI-Examples/python$ gramine-direct ./python scripts/helloworld.py
[P1:T1:python3.8] error: Internal memory fault at 0x00000000 (0x7ffff3f70793, VMID = 1, TID = 1)
```
It fails just after epoll_wait.
```
[P1:T1:python3.8] trace: ---- close(4) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/selector_events.py", 0x28b51da27b10) = 0x0
[P1:T1:python3.8] trace: ---- stat("/usr/lib/python3.8/asyncio/__pycache__", 0x28b51da27690) = 0x0
[P1:T1:python3.8] trace: ---- openat(AT_FDCWD, "/usr/lib/python3.8/asyncio/__pycache__/selector_events.cpython-38.pyc.44758334141104", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0644) = -13
[P1:T1:python3.8] trace: ---- clock_gettime(0, 0x28b51da2c5b0) = 0x0
[P1:T1:python3.8] trace: ---- epoll_create1(524288) = 0x4
[P1:T1:python3.8] trace: ---- clock_gettime(0, 0x28b51da2cc90) = 0x0
[P1:T1:python3.8] trace: ---- clock_gettime(1, 0x28b51da2cde0) = 0x0
[P1:T1:python3.8] trace: ---- epoll_wait(4, 0x28b51c91f320, 1023, 0) ...
[P1:T1:python3.8] error: Internal memory fault at 0x00000000 (0x7ffff3f70793, VMID = 1, TID = 1)

```

**Comment by dimakuv** [comment]
Thanks @deliciousdish. I debugged it a bit. Indeed, I can repro it easily on the latest master.

I'm running `GDB=1 gramine-direct ./python scripts/helloworld.py`.

The segfault happens here: https://github.com/gramineproject/gramine/blob/5fd2145632d81ff3d4af40899eb8f7fb2135e663/libos/src/sys/libos_epoll.c#L594

GDB extract:
```
(gdb) p items_count
$1 = 0
(gdb) p pal_handles
$2 = (PAL_HANDLE *) 0x0
(gdb) info threads
  Id   Target Id         Frame
* 1    LWP 1960708 "exe" do_epoll_wait (epfd=3, events=0x6c7627602240, maxevents=1023, timeout_ms=0)
    at ../libos/src/sys/libos_epoll.c:594
  2    LWP 1960712 "exe" gramine_raw_syscalls_code_begin () at ../pal/src/host/linux-common/arch/x86_64/syscall.S:28
```

So `pal_handles` array is somehow not allocated (or just the pointer nullified). At the same time, there is only one other thread running (`thread 2`), and it is the unrelated IPC helper thread. So this is not some data race.

Looking deeper at the code, I see that we never get into the IF clause at https://github.com/gramineproject/gramine/blob/5fd2145632d81ff3d4af40899eb8f7fb2135e663/libos/src/sys/libos_epoll.c#L530
```
(gdb) p arrays_len
$3 = 0
(gdb) p epoll->items_count
$5 = 0
```

Just for completeness, here is how the `epoll` object looks like (it's empty):
```
(gdb) p *epoll
$6 = {lock = {lock = 0x7ffff3ffbe50, owner = 1}, waiters = {first = 0x0}, items = {first = 0x0}, items_count = 0,
  last_returned_index = 18446744073709551615}
```

Btw, the arguments to `epoll_wait` syscall look like this:
```
#0  do_epoll_wait (epfd=3, events=0x6c7627602240, maxevents=1023, timeout_ms=0) at ../libos/src/sys/libos_epoll.c:594
#1  0x00007ffff3f4f578 in libos_syscall_epoll_wait (epfd=3, events=0x6c7627602240, maxevents=1023, timeout_ms=0)
    at ../libos/src/sys/libos_epoll.c:716
```

**TLDR:** Looks like our `do_epoll_wait()` code misses the corner case of `epoll->items_count == 0`. @boryspoplawski Am I correct? How can we fix this bug? Maybe we can simply add a corner case at the beginning of this func, since in this case epoll deteriorates to "sleep with a timeout".
